### PR TITLE
Bug 1146838 - Fix tab tray accessibility

### PR DESCRIPTION
**Pull request by dusek**
Make tab tray accessible again and enable tabs to be closed.

Patches donated by [A11Y LTD.](http://a11y.ltd.uk)

### CODE REVIEW

**Comment by AaronMT** [comment]
Ref bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1146838

File: Client/Frontend/Browser/TabTrayController.swift [hunk 1]
```diff
@@ -249,31 +253,40 @@ private class SwipeAnimator: NSObject {
...
             break
         }
     }
+
+    func close(#velocity: CGPoint, actualVelocity: CGFloat) {
+        // Calculate the edge to calculate distance from
+        let edgeX = velocity.x > 0 ? CGRectGetMaxX(self.container.frame) :
+            CGRectGetMinX(self.container.frame)
+        var distance
+        = (self.animatingView.frame.size.width / 2) + abs(self.animatingView.center.x - edgeX)
```
Comment (wesj) [inline Client/Frontend/Browser/TabTrayController.swift]: Keep this on one line

File: Client/Frontend/Browser/TabTrayController.swift [hunk 2]
```diff
@@ -249,31 +253,40 @@ private class SwipeAnimator: NSObject {
...
+
+        // Determine which way we need to travel
+        distance *= velocity.x > 0 ? 1 : -1
+
+        let timeStep = NSTimeInterval(abs(distance) / actualVelocity)
+        UIView.animateWithDuration(timeStep, animations: {
+            let animatedPosition
+            = CGPoint(x: self.animatingView.center.x + distance, y: self.animatingView.center.y)
```
Comment (wesj) [inline Client/Frontend/Browser/TabTrayController.swift]: Same here

File: Client/Frontend/Browser/TabTrayController.swift [hunk 3]
```diff
@@ -249,31 +253,40 @@ private class SwipeAnimator: NSObject {
...
+            self.animatingView.center = animatedPosition
+            }, completion: { finished in
```
Comment (wesj) [inline Client/Frontend/Browser/TabTrayController.swift]: XCode likes to indent these, but I usually push them back one level

File: Client/Frontend/Browser/TabTrayController.swift [hunk 4]
```diff
@@ -295,9 +308,6 @@ class TabTrayController: UIViewController, UITabBarDelegate, UITableViewDelegate
     var navBar: UINavigationBar!
 
     override func viewDidLoad() {
-        view.isAccessibilityElement = true
-        view.accessibilityLabel = NSLocalizedString("Tabs Tray", comment: "Accessibility label for the Tabs Tray view.")
```
Comment (wesj) [inline Client/Frontend/Browser/TabTrayController.swift]: We use this label for some KIF tests. Make sure they still pass.

File: Client/Frontend/Browser/TabTrayController.swift [hunk 5]
```diff
@@ -85,6 +85,10 @@ private class TabCell: UITableViewCell {
         setupFrames()
 
         self.animator.originalCenter = CGPoint(x: UIScreen.mainScreen().bounds.width / 2, y: TabTrayControllerUX.CellHeight / 2)
+
+        self.accessibilityCustomActions = [
+            UIAccessibilityCustomAction(name: NSLocalizedString("Close", comment: ""), target: self.animator, selector: "closeWithoutGesture")
```
Comment (wesj) [inline Client/Frontend/Browser/TabTrayController.swift]: Add a comment. Also, we always name selectors starting with "SEL" so they're easy to identify.

**Comment by dusek** [comment]
Fixed and interactively rebased to master with fixups. Please review.

**Comment by dusek** [comment]
(just fixed tabs->spaces in one of the commits now)

**Comment by dusek** [comment]
(now just removed @objc on the selector as it is not needed)

**Comment by dusek** [comment]
(and just added it back as it is really needed; should not code that late in the night :-)

**Comment by dusek** [comment]
I rebased on latest master, also added one (last) commit given there is now a close button on each tab.

**Comment by dusek** [comment]
I was also thinking that now there is a close button that closes tab without any animation. I originally thought if this would happen, it could use the animated close that I am using in the accessibility action. Or the other way: the accessibility close action could use the non-animated close. But I did not make any changes here to not complicate things at this point (i.e. delay merging of this).

**Comment by dusek** [comment]
Again rebased this, fixing conflicts. Please review again.

**Comment by splewako** [comment]
Please reference bug numbers in commit messages in the future, thanks.

**Comment by dusek** [comment]
@splewako sorry, will be more diligent about this the next time.

**Pull request merged**